ftw/tokenauth/oauth2/browser/oauth2_token.py

```diff
@@ -247,11 +254,36 @@ def validate_jwt_grant(self, assertion, service_key):
         # https://tools.ietf.org/html/rfc7523#section-3
         processor = JWTBearerGrantProcessor(self.plugin.get_token_uri())
         try:
-            processor.verify(assertion, service_key)
+            return processor.verify(assertion, service_key)
         except (VerificationError, InvalidTokenError) as exc:
             raise InvalidGrant(str(exc))
 
-    def issue_access_token(self, service_key):
+    def verified_subject(self, claimset, service_key):
+        """Verify the claim's subject and return it.
+
+        If the subject does not match the userid of the service_key, the
+        'Impersonate user' permission is required.
+        """
+        subject = claimset['sub']
+        actor = service_key['user_id']
+        if subject != actor:
+            # Check if actor is allowed to impersonate
+            uf = getToolByName(self.context, 'acl_users')
+            user = uf.getUserById(actor)
+            if not user:
+                raise(InvalidGrant('Service key user not found.'))
+            user = user.__of__(uf)
+            old_security_manager = getSecurityManager()
+            newSecurityManager(self.request, user)
+            allow_impersonation = getSecurityManager().checkPermission(
+                ImpersonateUser, self.context)
+            setSecurityManager(old_security_manager)
```

Comment:
I would wrap that in a try/finally block in order to ensure that the old security manager is always applied, even when an exception happens.

Although `checkPermission` probably won't raise an exception I find it important because it is good practice to use a try/finally block when fiddling with security managers and such code is prone to copy/paste where it cause damage..

```python
old_security_manager = getSecurityManager()
newSecurityManager(self.request, user)
try:
     allow_impersonation = getSecurityManager().checkPermission(ImpersonateUser, self.context)
finally:
    setSecurityManager(old_security_manager)
```

Reply:
Good point. Amended.